--- ../src-base/minecraft/net/minecraft/advancements/AdvancementTreeNode.java
+++ ../src-work/minecraft/net/minecraft/advancements/AdvancementTreeNode.java
@@ -1,6 +1,8 @@
 package net.minecraft.advancements;
 
 import com.google.common.collect.Lists;
+
+import java.util.Collections;
 import java.util.List;
 import javax.annotation.Nullable;
 
@@ -10,7 +12,7 @@
     private final AdvancementTreeNode parent;
     private final AdvancementTreeNode sibling;
     private final int index;
-    private final List<AdvancementTreeNode> children = Lists.<AdvancementTreeNode>newArrayList();
+    private final List<AdvancementTreeNode> children = Collections.synchronizedList(Lists.<AdvancementTreeNode>newArrayList());
     private AdvancementTreeNode ancestor;
     private AdvancementTreeNode thread;
     private int x;
