--- ../src-base/minecraft/net/minecraft/advancements/AdvancementProgress.java
+++ ../src-work/minecraft/net/minecraft/advancements/AdvancementProgress.java
@@ -17,6 +17,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
 import javax.annotation.Nullable;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.util.JsonUtils;
@@ -25,7 +26,7 @@
 
 public class AdvancementProgress implements Comparable<AdvancementProgress>
 {
-    private final Map<String, CriterionProgress> criteria = Maps.<String, CriterionProgress>newHashMap();
+    private final Map<String, CriterionProgress> criteria = new ConcurrentHashMap<>();
     private String[][] requirements = new String[0][];
 
     public void update(Map<String, Criterion> criteriaIn, String[][] requirements)
