--- ../src-base/minecraft/net/minecraft/entity/passive/EntityPig.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityPig.java
@@ -1,6 +1,8 @@
 package net.minecraft.entity.passive;
 
 import com.google.common.collect.Sets;
+
+import java.util.Collections;
 import java.util.Set;
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
@@ -37,12 +39,13 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityPig extends EntityAnimal
 {
     private static final DataParameter<Boolean> SADDLED = EntityDataManager.<Boolean>createKey(EntityPig.class, DataSerializers.BOOLEAN);
     private static final DataParameter<Integer> BOOST_TIME = EntityDataManager.<Integer>createKey(EntityPig.class, DataSerializers.VARINT);
-    private static final Set<Item> TEMPTATION_ITEMS = Sets.newHashSet(Items.CARROT, Items.POTATO, Items.BEETROOT);
+    private static final Set<Item> TEMPTATION_ITEMS = Collections.synchronizedSet(Sets.newHashSet(Items.CARROT, Items.POTATO, Items.BEETROOT));
     private boolean boosting;
     private int boostTime;
     private int totalBoostTime;
@@ -188,7 +191,7 @@
 
     public void onDeath(DamageSource cause)
     {
-        super.onDeath(cause);
+        // super.onDeath(cause); // CraftBukkit - Moved to end
 
         if (!this.world.isRemote)
         {
@@ -197,6 +200,7 @@
                 this.dropItem(Items.SADDLE, 1);
             }
         }
+        super.onDeath(cause);
     }
 
     @Nullable
@@ -226,7 +230,11 @@
     {
         if (!this.world.isRemote && !this.isDead)
         {
+            if (lightningBolt == null) lightningBolt = new EntityLightningBolt(this.world, this.posX, this.posY, this.posZ, true); // CatServer - if null, create EntityLightningBolt for Bukkit
             EntityPigZombie entitypigzombie = new EntityPigZombie(this.world);
+            if (CraftEventFactory.callPigZapEvent(this, lightningBolt, entitypigzombie).isCancelled()) {
+                return;
+            }
             entitypigzombie.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
             entitypigzombie.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, this.rotationPitch);
             entitypigzombie.setNoAI(this.isAIDisabled());
@@ -237,7 +245,7 @@
                 entitypigzombie.setAlwaysRenderNameTag(this.getAlwaysRenderNameTag());
             }
 
-            this.world.spawnEntity(entitypigzombie);
+            this.world.addEntity(entitypigzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
             this.setDead();
         }
     }
