--- ../src-base/minecraft/net/minecraft/pathfinding/Path.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/Path.java
@@ -7,6 +7,8 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+import java.util.concurrent.atomic.AtomicInteger;
+
 public class Path
 {
     private final PathPoint[] points;
@@ -14,18 +16,18 @@
     private PathPoint[] closedSet = new PathPoint[0];
     @SideOnly(Side.CLIENT)
     private PathPoint target;
-    private int currentPathIndex;
-    private int pathLength;
+    private volatile int  currentPathIndex;
+    private volatile int pathLength;
 
     public Path(PathPoint[] pathpoints)
     {
         this.points = pathpoints;
-        this.pathLength = pathpoints.length;
+        this.pathLength=pathpoints.length;
     }
 
     public void incrementPathIndex()
     {
-        ++this.currentPathIndex;
+        --this.currentPathIndex;
     }
 
     public boolean isFinished()
@@ -56,7 +58,7 @@
 
     public void setCurrentPathLength(int length)
     {
-        this.pathLength = length;
+        this.pathLength=length;
     }
 
     public int getCurrentPathIndex()
@@ -66,7 +68,7 @@
 
     public void setCurrentPathIndex(int currentPathIndexIn)
     {
-        this.currentPathIndex = currentPathIndexIn;
+        this.currentPathIndex=currentPathIndexIn;
     }
 
     public Vec3d getVectorFromIndex(Entity entityIn, int index)
@@ -160,7 +162,7 @@
         path.openSet = apathpoint1;
         path.closedSet = apathpoint2;
         path.target = pathpoint;
-        path.currentPathIndex = i;
+        path.currentPathIndex=i;
         return path;
     }
 }
