--- ../src-base/minecraft/net/minecraft/world/border/WorldBorder.java
+++ ../src-work/minecraft/net/minecraft/world/border/WorldBorder.java
@@ -1,17 +1,20 @@
 package net.minecraft.world.border;
 
 import com.google.common.collect.Lists;
+
+import java.util.Collections;
 import java.util.List;
 import net.minecraft.entity.Entity;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class WorldBorder
 {
-    private final List<IBorderListener> listeners = Lists.<IBorderListener>newArrayList();
+    private final List<IBorderListener> listeners = Lists.newArrayList();
     private double centerX;
     private double centerZ;
     private double startDiameter = 6.0E7D;
@@ -24,6 +27,9 @@
     private int warningTime;
     private int warningDistance;
 
+    // TODO: Check if it's initialized properly
+    public WorldServer world;
+
     public WorldBorder()
     {
         this.endDiameter = this.startDiameter;
@@ -206,6 +212,7 @@
 
     public void addListener(IBorderListener listener)
     {
+        if (listeners.contains(listener)) return;
         this.listeners.add(listener);
     }
 
@@ -284,4 +291,9 @@
             iborderlistener.onWarningDistanceChanged(this, warningDistance);
         }
     }
+
+    public void removeListener(IBorderListener listener)
+    {
+        this.listeners.remove(listener);
+    }
 }
